xdc-plugin/xdc.cc

```diff
@@ -268,13 +284,69 @@ struct SetProperty : public Pass {
 
 			// Set the parameter on the cell connected to the selected port
 			for (auto connection : cell->connections_) {
-				RTLIL::SigSpec cell_signals = connection.second;
-				if (!strcmp(log_signal(cell_signals), port_id.c_str())) {
+				RTLIL::SigSpec cell_signal = connection.second;
+				if (is_signal_port(cell_signal, port_name)) {
 					cell->setParam(parameter_id, RTLIL::Const(value));
 					log("Setting parameter %s to value %s on cell %s \n", parameter_id.c_str(), value.c_str(), cell_obj.first.c_str());
 				}
 			}
 		}
+		log("\n");
+	}
+
+	// Search module's connections for the specified destination port
+	// and traverse from the specified destination wire to the source wire
+	void traverse_wire(std::string& port_name, RTLIL::Module* module) {
+		auto port_signal = extract_signal(port_name);
+		std::string signal_name(port_signal.first);
+		int port_bit = port_signal.second;
+		for (auto connection : module->connections_) {
+			auto dst_sig = connection.first;
+			auto src_sig = connection.second;
+			if (dst_sig.is_chunk()) {
+				auto chunk = dst_sig.as_chunk();
+				if (chunk.wire) {
+					if (chunk.wire->name != RTLIL::IdString(RTLIL::escape_id(signal_name))) {
+						continue;
+					}
+					if (port_bit < chunk.offset || port_bit > chunk.width) {
+						continue;
+					}
+					auto src_wires = src_sig.to_sigbit_vector();
+					auto src_wire_sigbit = src_wires.at(port_bit - chunk.offset);
+					if (src_wire_sigbit.wire) {
+						port_name = src_wires.at(port_bit - chunk.offset).wire->name.str();
+						if (src_wire_sigbit.offset > 0) {
+							port_name += "[" + std::to_string(src_wire_sigbit.offset) + "]";
+						}
+						return;
+					}
+				}
+			}
+		}
+	}
+
+	// Extract signal name and port bit information from port name
+	std::pair<std::string, int> extract_signal(const std::string& port_name) {
+		char port[128];
```

Comment:
It seems weird to be mixing fixed length 128 byte stack strings with std::string.  Can you explain the reasoning why 128 characters is enough?  Why use a stack string over an allocated std::string?

Reply:
This was because I used sscanf. Otherwise I would go with allocated std::string. I guess using stringstream could be an alternative here.

Reply:
You can actually use std::string with sscanf, and because you know the size of the input string, you can safely allocate the maximum extract of port_name.size(), etc.

Reply:
Done